File: core/src/adictive/games/systems/FloorEffectSystem.java
package adictive.games.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.EntityListener;
import com.badlogic.ashley.core.EntitySystem;
import com.badlogic.gdx.utils.Array;

import adictive.games.components.MovementComponent;
import adictive.games.components.TransformComponent;
import adictive.games.utils.Families;


public class FloorEffectSystem extends EntitySystem implements Reseteable {
    private final Array<Entity> players = new Array<>();
    private final ComponentMapper<TransformComponent> tm;
    private final ComponentMapper<MovementComponent> mm;

    public FloorEffectSystem() {
        tm = ComponentMapper.getFor(TransformComponent.class);
        mm = ComponentMapper.getFor(MovementComponent.class);
    }

    @Override
    public void update(float deltaTime) {

    }

    @Override
    public void reset() {
        players.clear();
    }

    @Override
    public void addedToEngine(Engine engine) {
        super.addedToEngine(engine);
        engine.addEntityListener(
                Families.PLAYER,
                new EntityListener() {
                    @Override
                    public void entityAdded(Entity entity) {
                        players.add(entity);
                    }

                    @Override
                    public void entityRemoved(Entity entity) {
                        players.removeValue(entity, true);
                    }
                }
        );
    }
}

File: core/src/adictive/games/systems/FloorEffectSystem.kt
package adictive.games.systems

import adictive.games.SquareWorld
import adictive.games.TiledMap
import adictive.games.components.FloorEffectComponent
import adictive.games.components.MovementComponent
import adictive.games.components.TransformComponent
import adictive.games.utils.Families
import com.badlogic.ashley.core.*
import com.badlogic.gdx.utils.Array


class FloorEffectSystem(private val world: SquareWorld) : EntitySystem(), Reseteable {
    private val players = Array<Entity>()
    private val tm: ComponentMapper<TransformComponent>
    private val mm: ComponentMapper<MovementComponent>

    init {
        tm = ComponentMapper.getFor(TransformComponent::class.java)
        mm = ComponentMapper.getFor(MovementComponent::class.java)
    }

    override fun update(deltaTime: Float) {
        for (player in players) {
            val mc = mm.get(player)
            val tc = tm.get(player)

            val x = (tc.pos.x + tc.size.x / 2).toInt()
            val y = (tc.pos.y + tc.size.y / 2).toInt()

            val floor = world.tiledMap.getEntity(x, y, TiledMap.FLOOR)

            if (floor == null) {
                mc.friction = NORMAL_FRICTION
            } else if (floor.flags == FloorEffectComponent.ICE.toInt()) {
                mc.friction = ICE_FRICTION
            } else if (floor.flags == FloorEffectComponent.DIRT.toInt()) {
                mc.friction = DIRT_FRICTION
            } else if (floor.flags == FloorEffectComponent.ACCELERATOR.toInt()) {
                mc.friction = ACCELERATOR_FRICTION
            }
        }
    }

    override fun reset() {
        players.clear()
    }

    override fun addedToEngine(engine: Engine?) {
        super.addedToEngine(engine)
        engine!!.addEntityListener(
                Families.PLAYER,
                object : EntityListener {
                    override fun entityAdded(entity: Entity) {
                        players.add(entity)
                    }

                    override fun entityRemoved(entity: Entity) {
                        players.removeValue(entity, true)
                    }
                }
        )
    }

    companion object {

        private val ICE_FRICTION = 0f
        private val DIRT_FRICTION = 1.8f
        private val NORMAL_FRICTION = 0.2f
        private val ACCELERATOR_FRICTION = -11f
    }
}